### PR TITLE
Cosmoflow parallel io

### PR DESCRIPTION
**Pull request by enmccarthy**
draft pull request for cosmo hdf5 data reader

### CODE REVIEW

File: src/data_readers/data_reader_hdf5.cpp [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+        
+        //todo add error checking
+        H5Sselect_hyperslab(filespace, H5S_SELECT_SET, offset, NULL, count, dims_local);
+        H5Dread(h_data, H5T_NATIVE_SHORT, memspace, filespace, H5P_DEFAULT, data_out);
+    }
+
+    void hdf5_reader::load() {
+       
+        std::string dirpath = get_file_dir();    
+        std::vector<std::string> file_list = get_filenames(dirpath);
```
Comment (naoyam) [inline src/data_readers/data_reader_hdf5.cpp]: You probably need to update this part as you're now using `glob` to grab a list of all sample files.

File: src/data_readers/data_reader_hdf5.cpp [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        //m_transform_pipeline.apply(pixel_col, dims);
+        prof_region_end("fetch_datum", false);
+        return true;
+    }
+    //get from a cached response
+    bool hdf5_reader::fetch_response(Mat& Y, int data_id, int mb_idx) {
+        prof_region_begin("fetch_response", prof_colors[0], false);
+        Mat Y_v = El::View(Y, El::IR(0, Y.Height()), El::IR(mb_idx, mb_idx+1));
+        //TODO: possibly 4 tho, python tells me its float64
+        std::memcpy(Y_v.Buffer(), &m_all_responses,
```
Comment (naoyam) [inline src/data_readers/data_reader_hdf5.cpp]: Where is `m_all_responses` defined? 

I believe it should exist in `hdf5_reader` class. Make it `float`, not `double`.

File: src/proto/init_image_data_readers.cpp [hunk 3]
```diff
@@ -290,6 +291,8 @@ void init_org_image_data_reader(const lbann_data::Reader& pb_readme, const bool
   } else if (name == "cifar10") {
     reader = new cifar10_reader(shuffle);
     if (master) std::cout << "cifar10_reader is set" << std::endl;
+  } else if (name == "par_hdf5") {
```
Comment (naoyam) [inline src/proto/init_image_data_readers.cpp]: Do we still need this?

File: src/data_readers/data_reader_hdf5.cpp [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+        prof_region_begin("fetch_datum", prof_colors[0], false);
+        int nprocs;
+        MPI_Comm_size(MPI_COMM_WORLD, &nprocs); 
+        if ((nprocs%dc::get_number_of_io_partitions()) !=0) {
+            std::cerr<<"if other things have not been changed in the code this will not work for anything other than 4 procs a node \n";
+        }
+        int world_rank = get_rank_in_world();
+        double start_file = MPI_Wtime();
+        //TODO: do I need this mod --> will world rank/paritions ever be greater than the number of files??
+	    auto file = m_file_list[((world_rank/dc::get_number_of_io_partitions())+nux)%(m_file_list.size())];
```
Comment (naoyam) [inline src/data_readers/data_reader_hdf5.cpp]: `data_id` is the index of the sample to read. no need to calculate here. you can assume it's the same with a group of `dc::get_number_of_io_partitions()` procs.

File: src/data_readers/data_reader_hdf5.cpp [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+        prof_region_begin("fetch_datum", prof_colors[0], false);
+        int nprocs;
+        MPI_Comm_size(MPI_COMM_WORLD, &nprocs); 
+        if ((nprocs%dc::get_number_of_io_partitions()) !=0) {
+            std::cerr<<"if other things have not been changed in the code this will not work for anything other than 4 procs a node \n";
+        }
+        int world_rank = get_rank_in_world();
+        double start_file = MPI_Wtime();
+        //TODO: does this need to be divided by the num procs
+	    auto file = m_file_paths[data_id/dc::get_number_of_io_partitions()];
```
Comment (naoyam) [inline src/data_readers/data_reader_hdf5.cpp]: No need to divide.

File: src/data_readers/data_reader_hdf5.cpp [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+        DataType *dest = X_v.Buffer();
+        //TODO Add a reference to X mat so only one read
+	    //will this work ? ?
+        read_hdf5(h_data, filespace, world_rank, HDF5_KEY_DATA, dims, dest);
+        //close data set
+        H5Dclose(h_data);
+        if (m_has_responses) {
+            h_data = H5Dopen(h_file, HDF5_KEY_RESPONSES.c_str(), H5P_DEFAULT);
+            H5Dread(h_data, H5T_NATIVE_DOUBLE, H5S_ALL, H5S_ALL, H5P_DEFAULT, m_all_responses);
```
Comment (naoyam) [inline src/data_readers/data_reader_hdf5.cpp]: Since the 4 response values are stored in `double`, they need to be converted to `float`. Do something like this:
```
double responses[4];
H5Dread(h_data, ..., responses);
for (int i = 0; i < 4; ++i) {
  m_all_responses[i] = static_cast<float>(responses[i]);
}```

File: src/data_readers/data_reader_hdf5.cpp [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+            
+        if (h_data < 0) {
+            throw lbann_exception(std::string{} + __FILE__ + " " + std::to_string(__LINE__) +
+                                        " hdf5_reader::load() - can't find hdf5 key : " + HDF5_KEY_DATA);
+        }
+
+        //TODO: add the int 16 stuff
+	    //TODO: change the indexing 
+	    // check if mb_idx needs to be changed to not be hard coded
+	    int adj_mb_idx = mb_idx+(world_rank%dc::get_number_of_io_partitions());
```
Comment (naoyam) [inline src/data_readers/data_reader_hdf5.cpp]: Just `X.Buffer()` is fine for now. We only care about Cosmoflow, where we only have one local sample. No need to do column indexing.

File: src/utils/threads/thread_pool.cpp [hunk 8]
```diff
@@ -132,12 +132,13 @@ void thread_pool::do_thread_work_()
...
-    std::cerr << "error in pthread_setaffinity_np, error="
-              << error << std::endl;
-  }
+ // auto error = pthread_setaffinity_np(pthread_self(),
+   //                                   sizeof(cpu_set_t), &cpu_set);
+    //if (error != 0) {
+      //  std::cerr << "error in pthread_setaffinity_np, error="
+        //     << error << std::endl;
+        
+    //}
```
Comment (naoyam) [inline src/utils/threads/thread_pool.cpp]: Can you remove this modification? Or, is it necessary?

File: src/utils/lbann_library.cpp [hunk 9]
```diff
@@ -287,7 +282,6 @@ std::unique_ptr<model> build_model_from_prototext(
       r.second->setup_data_store(pb_model->mini_batch_size());
     }
   }
-
```
Comment (naoyam) [inline src/utils/lbann_library.cpp]: This is unrelated.

File: src/utils/lbann_library.cpp [hunk 10]
```diff
@@ -277,7 +273,6 @@ std::unique_ptr<model> build_model_from_prototext(
 
   // Setup models
   ret_model->setup();
-
```
Comment (naoyam) [inline src/utils/lbann_library.cpp]: Unrelated.

File: src/utils/lbann_library.cpp [hunk 11]
```diff
@@ -268,7 +265,6 @@ std::unique_ptr<model> build_model_from_prototext(
       }
     }
   }
-
```
Comment (naoyam) [inline src/utils/lbann_library.cpp]: Unrelated

File: src/utils/lbann_library.cpp [hunk 12]
```diff
@@ -216,6 +214,7 @@ std::unique_ptr<model> build_model_from_prototext(
   std::map<execution_mode, generic_data_reader *> data_readers;
   bool is_shared_training_data_reader = pb_model->shareable_training_data_reader();
   bool is_shared_testing_data_reader = pb_model->shareable_testing_data_reader();
+
```
Comment (naoyam) [inline src/utils/lbann_library.cpp]: Unrelated

File: src/proto/proto_common.cpp [hunk 13]
```diff
@@ -414,10 +423,12 @@ void init_data_readers(
     reader->set_master(master);
 
     reader->load();
-
+
```
Comment (naoyam) [inline src/proto/proto_common.cpp]: Unrelated.

File: model_zoo/vision/resnet.py [hunk 14]
```diff
@@ -194,12 +193,9 @@
...
     kwargs['lbann_args'] = (
-        '--data_filedir_train={} --data_filename_train={} '
-        '--data_filedir_test={} --data_filename_test={}'
-        .format(imagenet_dir(data_set='train', num_classes=classes),
-                imagenet_labels(data_set='train', num_classes=classes),
-                imagenet_dir(data_set='val', num_classes=classes),
-                imagenet_labels(data_set='val', num_classes=classes)))
+        '--data_filedir_train=/p/gpfs1/brainusr/datasets/cosmoflow/cosmoUniverse_2019_05_4parE/hdf5/21688988/ --data_filename_train=no'
+        '--data_filedir_test=/p/gpfs1/brainusr/datasets/cosmoflow/cosmoUniverse_2019_05_4parE/hdf5/21688988/ --data_filename_test=no'
+        )
```
Comment (naoyam) [inline model_zoo/vision/resnet.py]: resnet.py has nothing to do with this datareader.

File: src/data_readers/data_reader_hdf5.cpp [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+        // offset of the y dim for this rank
+        odd_offset = (yPerNode+1)*(dims[1]%yPerNode);
+        offset[1] = odd_offset + (yPerNode*(spatial_offset/ylines));
+      }   
+    } else {
+      offset[1] = 0;
+    }
+
+    if (zlines > 1) {
+      offset[2] = zPerNode*spatial_offset;
```
Comment (naoyam) [inline src/data_readers/data_reader_hdf5.cpp]: We should use the transposed data, so this is incorrect.

Comment (naoyam) [reply src/data_readers/data_reader_hdf5.cpp]: `offset[1]` should be set.

File: src/data_readers/data_reader_hdf5.cpp [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+    int xPerNode = dims[0]/xlines;
+    int yPerNode = dims[1]/ylines;
+    int zPerNode = dims[2]/zlines;
+    int cPerNode = dims[3]/channellines;
+    // offset in each dimension
+    hsize_t offset[4];
+    // how many times the pattern should repeat in the hyperslab
+    hsize_t count[4] = {1,1,1,1};
+    // local dimensions aka the dimensions of the slab we will read in
+    hsize_t dims_local[4] = {xPerNode, yPerNode, zPerNode, cPerNode};
```
Comment (naoyam) [inline src/data_readers/data_reader_hdf5.cpp]: the ordering is incorrect as we use the transposed dataset.

File: src/data_readers/data_reader_hdf5.cpp [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+////
+///////////////////////////////////////////////////////////////////////////////////
+#include "lbann/data_readers/data_reader_hdf5.hpp"
+#include "lbann/utils/profiling.hpp"
+#include <cstdio>
+#include <string>
+#include <fstream>
+#include <unordered_set>
+#include <iostream>
+#include "dirent.h"
```
Comment (naoyam) [inline src/data_readers/data_reader_hdf5.cpp]: `<dirent.h>`

File: src/data_readers/data_reader_hdf5.cpp [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+        // offset of this x dim for this rank;
+        // in most cases odd_offset will be 0
+        odd_offset = (xPerNode+1)*(dims[0]%xPerNode);
+        offset[0] = odd_offset +(xPerNode*(spatial_offset%xlines)); 
+      }
+    } else {
+      offset[0] = 0;
+    }
+
+    if (ylines > 1) {
```
Comment (naoyam) [inline src/data_readers/data_reader_hdf5.cpp]: We don't need this part of the code, so remove it for now.

File: src/data_readers/data_reader_hdf5.cpp [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+  bool hdf5_reader::fetch_label(Mat& Y, int data_id, int mb_idx) {
+    return true;
+  }
+  bool hdf5_reader::fetch_datum(Mat& X, int data_id, int mb_idx) {
+    prof_region_begin("fetch_datum", prof_colors[0], false);
+    int nprocs;
+    MPI_Comm_size(MPI_COMM_WORLD, &nprocs); 
+    if ((nprocs%dc::get_number_of_io_partitions()) !=0) {
+      std::cerr<<"nprocs should be divisible by num of io partitions otherwise this wont work \n";
+    }
```
Comment (naoyam) [inline src/data_readers/data_reader_hdf5.cpp]: Let's not do this every time a datum is read. It can be just done once, perhaps in `load()`.

File: src/data_readers/data_reader_hdf5.cpp [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+#include <string>
+#include <fstream>
+#include <unordered_set>
+#include <iostream>
+#include "dirent.h"
+#include <cstring>
+#include "lbann/utils/distconv.hpp"
+namespace lbann {
+  const std::string hdf5_reader::HDF5_KEY_DATA = "full";
+  const std::string hdf5_reader::HDF5_KEY_RESPONSES = "redshifts";
```
Comment (naoyam) [inline src/data_readers/data_reader_hdf5.cpp]: The name of the response parameters is actually `physPar`. I may have told you it's `redshifts`, sorry.

File: src/data_readers/data_reader_hdf5.cpp [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+    Mat X_v = El::View(X, El::IR(0,X.Height()), El::IR(mb_idx, adj_mb_idx+1));
+
+    DataType *dest = X_v.Buffer();
+    read_hdf5(h_data, filespace, world_rank, HDF5_KEY_DATA, dims, dest);
+    //close data set
+    H5Dclose(h_data);
+    if (m_has_responses) {
+      h_data = H5Dopen(h_file, HDF5_KEY_RESPONSES.c_str(), H5P_DEFAULT);
+      double responses[4];
+      H5Dread(h_data, H5T_NATIVE_DOUBLE, H5S_ALL, H5S_ALL, H5P_DEFAULT, responses);
```
Comment (naoyam) [inline src/data_readers/data_reader_hdf5.cpp]: The response parameters are stored as float.

File: src/data_readers/data_reader_hdf5.cpp [hunk 22]
```diff
@@ -0,0 +1,227 @@
...
+      for (int i = 0; i < 4; ++i) {
+        m_all_responses[i] = static_cast<float>(responses[i]);
+      }
+      H5Dclose(h_data);
+    }
+    H5Fclose(h_file);
+    double end_file= MPI_Wtime();
+    std::cerr<<"per file time " << end_file -start_file << "seconds to run. \n";
```
Comment (naoyam) [inline src/data_readers/data_reader_hdf5.cpp]: Remove these lines.

File: src/utils/threads/thread_pool.cpp (hunk 8 above)
Comment (enmccarthy) [reply src/utils/threads/thread_pool.cpp]: This was just more of a convenience, when I run on lassen this error prints for each rank. Tom said that is did not matter and that it was an issue of pthreads missing some setting with power9s. I can un-comment it.

Comment (naoyam) [reply src/utils/threads/thread_pool.cpp]: Yes, please uncomment it and minimize the other changes of this PR. Non essential changes should be recovered.

**Review by naoyam — APPROVED** [review]
